### ISSUE TITLE
is.na produced when calling sampling()

### ISSUE THREAD

**Issue by bob-carpenter** [stan-dev/rstan#188]
Andy Choi on stan-users reports the following warning messages which shouldn't be produced when calling the `sampling)` function:

```
Warning messages:
1: In is.na(x) : is.na() applied to non-(list or vector) of type 'closure'
2: In FUN(X[[i]], ...) : data with name time is not numeric and not used
```

```
N_d <- 20

forecast_error <- c(
            0.37733820,             0.34814320,             0.37601800, 
            0.39766186,             0.40671559,             0.37047201,
            0.32281526,             0.29462281,             0.30124360, 
            0.29665000,             0.30814859,             0.28484706, 
            0.26505298,             0.29368737,             0.28023730, 
            0.19997346,             0.17094374,             0.14203103, 
            0.12186337,             0.08997575
);

t1 <- stan_model ("t1.stan", model_name = "t1");



mkt.error.fit <- sampling (t1, 
    #EA.stanEngine.model.estMarketEff.fit,
                           data = list (N_d = N_d, forecast_error = forecast_error),
                           iter=10000, chains=4);
#print ("done");

> source ("t1.r")
```

Output:

```
COMPILING THE C++ CODE FOR MODEL 't1' NOW.

SAMPLING FOR MODEL 't1' NOW (CHAIN 1).

Chain 1, Iteration:    1 / 10000 [  0%]  (Warmup)
Informational Message: The current Metropolis proposal is about to be rejected because of the following issue:
Exception thrown at line 40:
stan::math::normal_log: Scale parameter is 0, but must be > 0!
If this warning occurs sporadically, such as for highly constrained variable types like covariance matrices, then the sampler is fine,
but if this warning occurs often then your model may be either severely ill-conditioned or misspecified.

...

Chain 4, Iteration: 10000 / 10000 [100%]  (Sampling)
#  Elapsed Time: 1.36565 seconds (Warm-up)
#                1.97616 seconds (Sampling)
#                3.34181 seconds (Total)

Warning messages:
1: In is.na(x) : is.na() applied to non-(list or vector) of type 'closure'
2: In FUN(X[[i]], ...) : data with name time is not numeric and not used
```

And here's `t1.stan` (minus some comments):

```
data {
  int <lower=2> N_d;       // number of discretization
  real forecast_error[N_d];
}

transformed data {
  real        log_unif;
  vector[N_d] time;

  log_unif <- -log(N_d);

  {
    real stepping;
    stepping <- 1.0/N_d;

    for (i in 1:N_d) {
      time[i] <- i * stepping;
    }
  }
}

parameters { 
  real<lower=0> rmse_static;
  real<lower=0> rmse_static_sd;

  real<upper=0> learn_rate;
  real<lower=0> learn_sd;
}

transformed parameters {
  vector[N_d] lp;
  lp <- rep_vector(log_unif, N_d);

  // special case, n=1
  lp[1] <- lp[1] + normal_log(forecast_error, learn_rate * time + rmse_static, learn_sd);


  for (s in 2:N_d) {
    int len_s;
    int len_l;

    len_s <- s-1;
    len_l <- N_d-s+1;

    {
      real forecast_error_i[len_s];
      real forecast_error_l[len_l];
      vector[len_l] time_l;

      forecast_error_i <- head (forecast_error, len_s);
      forecast_error_l <- tail (forecast_error, len_l);      
      time_l <- head (time, len_l);

      lp[s] <- lp[s] + 
    normal_log (forecast_error_i, rmse_static, rmse_static_sd) +
    normal_log (forecast_error_l, learn_rate * time_l + rmse_static, learn_sd);
    }
  }
}

model {
  rmse_static ~ normal (0, 0.5);
  rmse_static_sd ~ normal (0, 0.5);

  learn_rate ~ normal (0, 2);
  learn_sd ~ normal (0, 0.5);

  increment_log_prob(log_sum_exp(lp));
}
```

**Comment by bgoodri** [comment]
This is a weakness of `rstan:::parse_data` which does not distinguish between objects declared in the data block vs. objects that are declared in the transformed data block. In this case, `time` is declared in transformed data but is also an R function, so there is a warning when it picks that up and later applies `is.na()` to it.

**Comment by bgoodri** [comment]
Fixed in 07616f8
